File: backend/ansible_keys.py
"""
ansible_keys.py — The secrets an automation needs, and where they go (#586).

A playbook run needs credentials ShellMate does not otherwise hold: an
Azure client secret, an AWS access key, a Meraki API key, a vault password,
the passphrase on the runner's SSH key. They cannot live in the playbook,
they should not be typed into the run dialog every time, and they must not
be written to ``settings.json``.

So they live in ShellMate's encrypted vault under a name, and a run refers
to them **by that name**. An environment says "this run needs
``azure_secret`` as ``AZURE_CLIENT_SECRET``"; at the moment the run starts,
the value is fetched from the vault and travels with the job. Nothing is
stored on the runner, and nothing crosses the wire until a run actually
needs it.

Two honest limits, stated here because the panel states them too:

- **The value does reach the runner.** It has to: Ansible is what uses it.
  What this buys is that the secret is not in a playbook, not in a file on
  the container, and not in a shell history — and that it goes over the
  connection ShellMate already trusts rather than being pasted somewhere.
- **The value never comes back.** Listing keys returns names, kinds and
  when they were set. There is no endpoint that reads one; the only way
  out is a run. A key you cannot remember has to be replaced, which is the
  correct trade for a store nothing can exfiltrate through the API.

Metadata (the names, what each is for) is an ordinary JSON file; only the
values are in the vault. That way the panel can list keys with the vault
locked, and say plainly which of them are currently unreadable.
"""

import logging
import re
import time
import uuid

from backend import jsonfile, paths

logger = logging.getLogger(__name__)


def _file():
    return paths.data_dir() / "ansible" / "keys.json"


class KeyError_(ValueError):
    """A key that cannot be stored as asked."""


#: How a key is delivered to the run. The distinction is not cosmetic:
#: an environment variable is visible to every task in the play, while an
#: extra var is a variable a playbook has to name to use.
DELIVERY = ("env", "extra_var")

#: What a key is for. Used only to group the list and to pick sensible
#: default variable names — nothing branches on it.
KINDS = {
    "generic": "Something else",
    "cloud": "A cloud API credential",
    "device": "A device or network credential",
    "vault": "An Ansible Vault password",
    "ssh": "An SSH key passphrase",
}

_NAME_RE = re.compile(r"^[a-z_][a-z0-9_]{0,63}$")

#: The vault key a stored value lives under. Prefixed so an Ansible key can
#: never collide with ``ansible_token`` or a provider's API key.
_VAULT_PREFIX = "ansible_key:"


def _vault():
    from backend.vault import vault

    return vault


def _load() -> list[dict]:
    return jsonfile.read(_file(), [], expect=list)


def keys() -> list[dict]:
    """
    Every key, without a single value.

    ``readable`` says whether the vault can currently produce the value.
    A locked vault is not an error here — the list is still worth showing,
    and a run that needs an unreadable key fails with that as its reason
    rather than with an empty variable.
    """
    rows = sorted(_load(), key=lambda k: k.get("name") or "")
    for row in rows:
        row.pop("value", None)                 # belt and braces; never stored here
        row["readable"] = bool(_read_value(row.get("name", "")))
    return rows


def _read_value(name: str) -> str:
    if not name:
        return ""
    try:
        return _vault().get(_VAULT_PREFIX + name, "") or ""
    except Exception:                                       # locked, or no vault
        return ""


def save_key(fields: dict) -> dict:
    """
    Store or update a key.

    An empty ``value`` on an existing key leaves the stored value alone, so
    the panel can rename a key or change where it is delivered without the
    user having to retype a secret they may not have.
    """
    name = str(fields.get("name") or "").strip()
    if not _NAME_RE.match(name):
        raise KeyError_(
            "A key name is lower case letters, digits and underscores, "
            "starting with a letter or underscore — it becomes a variable.")

    delivery = str(fields.get("delivery") or "env")
    if delivery not in DELIVERY:
        raise KeyError_("A key is delivered as an environment variable or an extra var.")

    kind = str(fields.get("kind") or "generic")
    if kind not in KINDS:
        raise KeyError_(f"'{kind}' is not a kind of key ShellMate knows.")

    target = str(fields.get("target") or "").strip()
    if not target:
        target = name.upper() if delivery == "env" else name
    if delivery == "env" and not re.match(r"^[A-Za-z_][A-Za-z0-9_]*$", target):
        raise KeyError_("An environment variable name holds letters, digits "
                        "and underscores, and cannot start with a digit.")
    if delivery == "extra_var" and not _NAME_RE.match(target):
        raise KeyError_("An extra var name is lower case letters, digits and "
                        "underscores.")

    value = str(fields.get("value") or "")
    entry_id = str(fields.get("id") or "")

    path = _file()
    with jsonfile.locked(path):
        rows = jsonfile.read(path, [], expect=list)
        existing = next((r for r in rows if r.get("id") == entry_id), None)
        clash = next((r for r in rows
                      if r.get("name") == name and r.get("id") != entry_id), None)
        if clash:
            raise KeyError_(f"There is already a key called '{name}'.")
        if not value and existing is None:
            raise KeyError_("A new key needs a value.")

        # A rename has to move the stored value, or the key stops resolving.
        old_name = (existing or {}).get("name", "")
        carried = ""
        if existing and old_name and old_name != name and not value:
            carried = _read_value(old_name)
            if not carried:
                raise KeyError_(
                    "The vault cannot read this key's value at the moment, so "
                    "renaming it would lose it. Unlock the vault, or set a new "
                    "value with the new name.")

        entry = {
            "id": entry_id or str(uuid.uuid4()),
            "name": name,
            "kind": kind,
            "delivery": delivery,
            "target": target,
            "description": str(fields.get("description") or "").strip(),
            "updated": time.time(),
        }
        if existing is not None:
            rows[rows.index(existing)] = entry
        else:
            rows.append(entry)

        if value or carried:
            _vault().set(_VAULT_PREFIX + name, value or carried)
            if old_name and old_name != name:
                try:
                    _vault().delete(_VAULT_PREFIX + old_name)
                except Exception:
                    logger.warning("Old Ansible key value left in the vault: %s", old_name)
        jsonfile.write(path, rows)

    logger.info("Ansible key saved: %s (%s as %s)", name, delivery, target)
    entry["readable"] = bool(_read_value(name))
    return entry


def delete_key(entry_id: str) -> bool:
    """Forget a key, value and all."""
    path = _file()
    with jsonfile.locked(path):
        rows = jsonfile.read(path, [], expect=list)
        gone = next((r for r in rows if r.get("id") == entry_id), None)
        if gone is None:
            return False
        jsonfile.write(path, [r for r in rows if r.get("id") != entry_id])
    try:
        _vault().delete(_VAULT_PREFIX + gone.get("name", ""))
    except Exception:
        logger.warning("Vault value survived a deleted Ansible key: %s", gone.get("name"))
    return True


def resolve(names: list[str]) -> tuple[dict, dict, list[str]]:
    """
    Turn key names into what a run actually carries.

    Returns the environment variables, the extra vars, and the names that
    could not be read. The caller stops on that third list rather than
    starting a run with a blank credential — Ansible's failure for an empty
    password is several screens away from the cause.
    """
    wanted = {n for n in (names or []) if n}
    env: dict[str, str] = {}
    extra: dict[str, str] = {}
    unreadable: list[str] = []
    for row in _load():
        if row.get("name") not in wanted:
            continue
        value = _read_value(row["name"])
        if not value:
            unreadable.append(row["name"])
            continue
        if row.get("delivery") == "extra_var":
            extra[row.get("target") or row["name"]] = value
        else:
            env[row.get("target") or row["name"].upper()] = value
    known = {r.get("name") for r in _load()}
    unreadable.extend(sorted(wanted - known))
    return env, extra, sorted(set(unreadable))


def count() -> int:
    try:
        return len(_load())
    except Exception:                                       # pragma: no cover
        return 0

File: frontend/js/ansible_keys.js
/**
 * ansible_keys.js — Credentials a run needs, held in the vault and sent only with a run that needs them (#586).
 *
 * `backend/ansible_keys.py` states the limits of this store in its own
 * module docstring, because they are the kind of thing a UI is tempted to
 * paper over. This file states them again, on screen, for the same reason:
 * a locked door with a sign reading "not actually locked" is worse than no
 * sign at all. Nothing below softens any of the three:
 *
 *   1. There is no way to read a stored value back. Listing returns names,
 *      kinds and where each is delivered — never the value. A key nobody
 *      remembers has to be replaced.
 *   2. The value does reach the runner. It has to — Ansible is what uses
 *      it. What the vault buys is that the secret is never in a playbook,
 *      never in a file on the container, and never in a shell history.
 *   3. While a run executes, the value sits in the runner container's
 *      process environment, and a playbook that prints it leaks it itself.
 *      `no_log: true` is the play author's job, not something reachable
 *      from here.
 *
 * The create/edit dialog is `shellmateDialog.form`, which already masks a
 * password field and gives it a reveal toggle — that toggle shows what was
 * just typed into the box, which is not the same thing as reading back what
 * is stored, and does not contradict point 1 above.
 */

(function () {
  'use strict';

  const view = window.ansibleView;
  if (!view) return;
  const { el, icon, clear, empty } = view;

  /** What the backend falls back to before `/api/ansible/keys` has answered. */
  const FALLBACK_KINDS = {
    generic: 'Something else',
    cloud: 'A cloud API credential',
    device: 'A device or network credential',
    vault: 'An Ansible Vault password',
    ssh: 'An SSH key passphrase',
  };

  function currentKinds() {
    const kinds = (view.state.keys || {}).kinds || {};
    return Object.keys(kinds).length ? kinds : FALLBACK_KINDS;
  }

  // -- Create / edit ---------------------------------------------------------

  async function openForm(entry) {
    const kinds = currentKinds();
    const kindOptions = Object.entries(kinds).map(([value, label]) => ({ value, label }));

    const fields = [
      { name: 'name', label: 'Name', value: entry ? entry.name : '', required: true,
        hint: 'Lower case letters, digits and underscores, starting with a '
            + 'letter or underscore — this is the name a run refers to.' },
      { name: 'kind', label: 'Kind', type: 'select',
        value: entry ? entry.kind : 'generic', options: kindOptions },
      { name: 'delivery', label: 'Delivered as', type: 'select',
        value: entry ? entry.delivery : 'env',
        options: [
          { value: 'env', label: 'Environment variable — visible to the whole play' },
          { value: 'extra_var', label: 'Extra var — a task has to name it' },
        ] },
      { name: 'target', label: 'Target name', value: entry ? entry.target : '',
        hint: 'Leave blank for the default: the name in capitals for an '
            + 'environment variable, or the name as it is for an extra var.' },
      { name: 'description', label: 'Description', value: entry ? entry.description : '' },
      { name: 'value', label: entry ? 'New value' : 'Value', type: 'password',
        required: !entry,
        hint: entry
          ? 'Leave this blank to keep the value already stored — nothing here '
          + 'can show you what that is, so this is the only way to leave it '
          + 'alone. Type a new one only to replace it.'
          : "Held in the encrypted vault, never written into a playbook or "
          + "into this list. It reaches the runner's own process environment "
          + "the moment a run needs it, and there is no way to read it back "
          + "afterwards." },
    ];

    const values = await window.shellmateDialog.form({
      title: entry ? `Edit ${entry.name}` : 'New key',
      fields,
      confirmLabel: 'Save',
    });
    if (!values) return;

    const payload = {
      id: entry ? entry.id : '',
      name: values.name,
      kind: values.kind,
      delivery: values.delivery,
      target: values.target,
      description: values.description,
      value: values.value,
    };

    try {
      await view.post('/api/ansible/keys', payload);
      await view.load();
    } catch (e) {
      view.toast(e.message, 'error');
    }
  }

  async function remove(entry) {
    const ok = await window.shellmateDialog.confirm({
      title: `Delete ${entry.name}?`,
      body: 'The stored value is deleted with it — this cannot be undone, and '
          + 'nothing can bring the value back.',
      danger: true,
      confirmLabel: 'Delete',
    });
    if (!ok) return;
    try {
      await view.del(`/api/ansible/keys/${encodeURIComponent(entry.id)}`);
      await view.load();
    } catch (e) {
      view.toast(e.message, 'error');
    }
  }

  // -- Rendering ---------------------------------------------------------------

  /**
   * The three limits, said once at the top rather than folded into a
   * tooltip somebody would have to go looking for.
   */
  function notice() {
    return el('div', { class: 'av-notice av-notice-info av-key-notice' }, [
      icon('info'),
      el('div', {}, [
        el('p', {}, [
          el('strong', { text: 'There is no way to see a stored value again. ' }),
          'Listing only ever returns names, kinds and where each is delivered. '
          + 'A key nobody remembers has to be replaced.',
        ]),
        el('p', {}, [
          el('strong', { text: 'The value does reach the runner. ' }),
          'It has to — Ansible is what uses it. What the vault buys is that '
          + 'the secret is never in a playbook, never in a file on the '
          + 'container, and never in a shell history.',
        ]),
        el('p', {}, [
          el('strong', { text: 'While a run executes, the value sits in the '
            + "runner container's process environment. " }),
          'A playbook that prints the variable leaks it itself — '
          + 'no_log: true is the play author’s job, and neither of these '
          + 'can be closed from ShellMate’s side.',
        ]),
      ]),
    ]);
  }

  function statusCell(entry) {
    if (entry.readable) {
      return el('span', { class: 'av-key-status av-pill av-pill-ok' },
        [icon('encrypted'), 'Readable']);
    }
    return el('div', {}, [
      el('span', { class: 'av-key-status av-pill av-pill-warn' },
        [icon('lock'), 'Unreadable']),
      el('p', { class: 'av-key-status-fix',
        text: 'This will stop a run that needs it. Unlock the vault, or set '
            + 'the value again.' }),
    ]);
  }

  function row(entry) {
    return el('tr', {}, [
      el('td', { class: 'av-key-name', text: entry.name }),
      el('td', {}, el('code', { class: 'av-key-target', text:
        `${entry.delivery === 'extra_var' ? 'extra var' : 'env'} ${entry.target}` })),
      el('td', { class: 'av-key-desc', title: entry.description || '',
        text: entry.description || '—' }),
      el('td', {}, statusCell(entry)),
      el('td', { class: 'av-row-actions' }, [
        el('button', { type: 'button', class: 'icon-btn', title: 'Edit',
          onclick: () => openForm(entry) }, icon('edit')),
        el('button', { type: 'button', class: 'icon-btn', title: 'Delete',
          onclick: () => remove(entry) }, icon('delete_forever')),
      ]),
    ]);
  }

  function kindBlock(label, rows) {
    return el('section', { class: 'av-block' }, [
      el('h4', { class: 'av-block-title', text: `${label} (${rows.length})` }),
      el('table', { class: 'av-table av-key-table' }, [
        el('thead', {}, el('tr', {}, [
          el('th', { text: 'Name' }), el('th', { text: 'Delivered as' }),
          el('th', { text: 'Description' }), el('th', { text: 'Status' }),
          el('th', { text: '' }),
        ])),
        el('tbody', {}, rows.map(row)),
      ]),
    ]);
  }

  function newButton(style) {
    return el('button', {
      type: 'button', class: style === 'primary' ? 'btn-primary' : 'btn-secondary',
      onclick: () => openForm(null),
    }, [icon('add'), 'New key']);
  }

  function render(state) {
    const body = document.getElementById('av-keys-body');
    if (!body) return;
    clear(body);
    body.appendChild(notice());

    const keys = ((state.keys || {}).keys) || [];
    const kinds = currentKinds();

    if (!keys.length) {
      body.appendChild(view.blank({
        icon: 'key',
        title: 'Credentials a run needs, held in the vault',
        lines: [
          'A key holds one credential a playbook needs and ShellMate does not '
          + 'otherwise have — a cloud API key, an Ansible Vault password. It '
          + 'lives in the encrypted vault under a name, and a run refers to it '
          + 'by that name; the value is fetched only as the run starts.',
          'Each key is delivered either as an environment variable, which '
          + 'every task can see, or as an extra var, which a playbook has to '
          + 'name to use. They are not the same thing.',
          'No value can ever be read back. Listing keys returns names and '
          + 'whether the vault can currently read them, and nothing returns a '
          + 'value — so a key you cannot remember has to be replaced.',
          'And the part worth being plain about: the value does reach the '
          + 'runner, because Ansible is what uses it. What the vault buys is '
          + 'that it is not in a playbook, not in a file on the container and '
          + 'not in a shell history. It is not end-to-end secrecy.',
        ],
        action: newButton('primary'),
      }));
      return;
    }

    body.appendChild(el('div', { class: 'av-key-toolbar' }, newButton('primary')));

    const grouped = new Set();
    Object.keys(kinds).forEach((kind) => {
      const rows = keys.filter((k) => k.kind === kind);
      if (!rows.length) return;
      grouped.add(kind);
      body.appendChild(kindBlock(kinds[kind], rows));
    });
    const strays = keys.filter((k) => !grouped.has(k.kind));
    if (strays.length) body.appendChild(kindBlock('Other', strays));
  }

  view.area('keys', {
    onShow: (state) => render(state),
    onData: (state) => { if (view.current === 'keys') render(state); },
  });
})();
